### ISSUE TITLE
Splitting langchaingo into sub-modules

### ISSUE THREAD

**Issue by eliben** [tmc/langchaingo#369]
Currently langchaingo is a single Go module (modulo the examples, which live in separate modules per example - consistently after #367 lands).

This means that users who are only interested in using langchaingo with a single backend LLM (say, OpenAI), will see their `go` commands pull in many dependencies they don't necessarily need. Even though the `go build` tool won't include them in the users' programs, this may have negative effects for - say - CI time. But it's important to hear what specific concerns users are having.

Splitting the repo into multiple modules is natural on the LLM provider level; e.g. `langchaingo` can be its own (top-level) module, and each LLM provider another module nested within the repo, importing the main `langchaingo` module for the common stuff. This way users can only import the modules they need and their dependencies.

This isn't free of cost, however. Managing multiple modules in a repo requires some care, especially around releases. In a way it's like each module living in its own separate repo, except that in a single repo they can share tooling and scripts. Also, with the existence of go workspaces, a `go.work` file can make local development much more pleasant.

One pre-requisite for this would be to start tagging actual releases of `langchaingo`. We can start with `0.1.0` and follow semver to increase the minor version as much as we need.

**Comment by eliben** [comment]
I just tested this by copying `examples/openai-completion-example` into a standalone directory with a new module, and ran `go mod tidy` to see what is pulled in:

```
$ go mod tidy
go: finding module for package github.com/tmc/langchaingo/llms/openai
go: finding module for package github.com/tmc/langchaingo/llms
go: found github.com/tmc/langchaingo/llms in github.com/tmc/langchaingo v0.0.0-20231122191601-2eb6f5408849
go: found github.com/tmc/langchaingo/llms/openai in github.com/tmc/langchaingo v0.0.0-20231122191601-2eb6f5408849
```

It's not that bad, actually! The Go tool is good about only pulling whichever dependencies are needed.

**Comment by tmc** [comment]
Is it my imagination or did it not used to be as smart with unused transitive dependencies?

**Comment by eliben** [comment]
Yes, this is module pruning which was shipped in Go 1.17 -- see https://go.dev/ref/mod#graph-pruning (there's a link there to a design doc with more details if you're interested)

**Comment by tmc** [comment]
I'd still like to find ways to keep the number of dependencies low and want to explore this more -- we could perhaps analyze what is bringing in the most and consider more targeted submodules there.

**Comment by tmc** [comment]
I'm experimenting with what having modules under ./contrib would look like here: https://github.com/tmc/langchaingo/tree/add-contrib

**Comment by eliben** [comment]
I think the `add-contrib` branch looks alright if you want to go in this direction. 

Just like with the examples, the dependency has to be always one way - the contrib modules depend on the main `langchaingo` module, not the other way around. And dependency management will have to be done similarly with bumping main module versions in all the `go.mod` when the main module has a new release. This creates an issue when you want to add a feature to the main module and immediately use it in a contrib module because there's no new tag yet; it can be tested with a `go.work` file locally, but CI may be unhappy on the PR.

**Comment by tmc** [comment]
Yeah, I’m thinking we either do some ci automation around go.work files.

Another option would be something that would auto-tag pre-release tags or
branches so that pushes to branches would become go-gettable.

Thoughts there?

Could do something like branch “pr-$n” gets created/updated when you open a
PR.

**Comment by eliben** [comment]
It's possible, but I don't think it makes sense to over-complicate this, honestly.

Most changes don't need to affect all modules at the same time; since `go.mod` versions don't get auto-bumped, things should work in between. E.g. if we make a change in main module `vN`, then want to use this change in a sub-module, we can have one commit updating the main module, then tag `vN`, then another commit updating the `go.mod` line in a contrib module to use `vN` and the code change.

Single commit/PR affecting both main and contrib modules should hopefully be rare.

I would recommend waiting with automation to see what actual issues we encounter in this repository, since each project is different in this respect.

**Comment by devinyf** [comment]
Just wondering... Is it a good idea to use go-plugins to manage third-party modules like llms, tools ??

**Comment by prestonvasquez** [comment]
What is that status of this? At the moment, it could be argued that taking a dependency on N+ storage drivers is not worth using this library. I would imagine a lot of users are simply copying the minimal amount of code required to create a receiver.

**Comment by ryancurrah** [comment]
@tmc I’m on a Developer Experience team and we’re evaluating langchaingo for internal use.

One concern from our dependency review is the repo being a single module with all providers included. Even with Go’s module graph pruning, our CI and supply-chain scanning still see the full module’s dependency surface, and we’ve been burned before by “one repo pulls everything” when a transitive dependency becomes a migration blocker (e.g., AWS SDK v1 → v2).

Is there appetite nowadays to split this into a small “core” module plus provider/tool/vectorstore submodules (similar to the earlier ./contrib approach)? Something like:
- github.com/tmc/langchaingo (core interfaces + shared utilities)
- github.com/tmc/langchaingo/llms/openai (and other providers) as separate modules

If that direction is interesting, we can help with:
- a concrete module layout proposal (what stays core vs contrib)
- release/tagging strategy for multiple modules
- CI/go.work workflow so changes spanning core + contrib stay painless
- PRs to implement the split and migrate a few providers as examples
